extract_token_from_browser.py

```diff
@@ -1,64 +1,176 @@
+        sys.exit(1)
+else:
+    # b'v10' is for Chromium, but not Chrome, it seems?
+    prefix =  b'v11'
+    cookie_d_query = (
+        "SELECT encrypted_value FROM cookies WHERE "
+        "host_key = '.slack.com' AND name = 'd'"
+    )
+    cookie_ds_query = (
+        "SELECT encrypted_value FROM cookies WHERE "
+        "host_key = '.slack.com' AND name = 'd-s'"
+    )
+    if not profile:
+        profile = "Default"
+
+    default_profile_path = browser_data.joinpath(profile)
+    cookies_path = default_profile_path.joinpath("Cookies")
+
+if not cookies_path:
+    cookies_path = default_profile_path.joinpath("cookies.sqlite")
+
+cookie_d_value = None
+cookie_ds_value = None
 try:
-    default_profile_path = next(firefox_path.glob("*.default-release"))
-except StopIteration:
-    print("Couldn't find the default profile for Firefox", file=sys.stderr)
-    sys.exit(1)
+    con = sqlite3.connect(f"file:{cookies_path}?immutable=1", uri=True)
+    cookie_d_value = con.execute(cookie_d_query).fetchone()[0]
+    cookie_ds_value = con.execute(cookie_ds_query).fetchone()[0]
+except TypeError:
+    if not cookie_d_value:
+        print("Couldn't find the 'd' cookie value", file=sys.stderr)
+        sys.exit(1)
+    pass
+finally:
+    con.close()
 
-cookies_path = default_profile_path.joinpath("cookies.sqlite")
-con = sqlite3.connect(f"file:{cookies_path}?immutable=1", uri=True)
-cookie_d_query = (
-    "SELECT value FROM moz_cookies WHERE host = '.slack.com' AND name = 'd'"
-)
-cookie_d_value = con.execute(cookie_d_query).fetchone()[0]
-cookie_ds_query = (
-    "SELECT value FROM moz_cookies WHERE host = '.slack.com' AND name = 'd-s'"
-)
-cookie_ds_values = con.execute(cookie_ds_query).fetchone()
-con.close()
+if args.browser in ["chrome", "chrome-beta"]:
+    bus = secretstorage.dbus_init()
+    collection = secretstorage.get_default_collection(bus)
+    for item in collection.get_all_items():
+        if item.get_label() == 'Chrome Safe Storage':
+            passwd = item.get_secret()
+            break
+    else:
+        raise Exception('Chrome password not found!')
 
-if cookie_ds_values:
-    cookie_value = f"d={cookie_d_value};d-s={cookie_ds_values[0]}"
+    salt = b'saltysalt'
+    length = 16
+    key = PBKDF2(passwd, salt, length, iterations)
+
+    cipher = AESCipher(key)
+    
+    if cookie_d_value[:3] == prefix:
+        cookie_d_value = cipher.decrypt(cookie_d_value[3:]).decode('utf8')
+
+    if cookie_ds_value and cookie_ds_value[:3] == prefix:
+       cookie_ds_value = cipher.decrypt(cookie_ds_value[3:]).decode('utf8')
+
+if cookie_ds_value:
+    cookie_value = f"d={cookie_d_value};d-s={cookie_ds_value}"
 else:
     cookie_value = cookie_d_value
 
 local_storage_path = default_profile_path.joinpath("webappsstore.sqlite")
-con = sqlite3.connect(f"file:{local_storage_path}?immutable=1", uri=True)
 local_storage_query = "SELECT value FROM webappsstore2 WHERE key = 'localConfig_v2'"
```

Comment:
I see you haven't changed this. Does this mean that it will always fail for Chrome? I assume Chrome doesn't have the same structure for local storage as Firefox?

Reply:
it wasn't failing for Chrome, just returning empty results

Reply:
Well, apparently this file exists, but it's empty for me at least. I don't know why Chrome has this empty file, but I'm pretty sure trying to read it is useless. Just getting the token and not the Cookies out of Chrome feels pretty pointless to me as you would still have to go to the browser to get out the tokens (and I'd argue that it's more work/less obvious how to get out the tokens than the cookie).

I checked a bit though and see that we can read out the token from Chromes local storage. It can be done like this (for Chromium, but Chrome is probably the same just with a different path):

```
import json
import plyvel

db = plyvel.DB('/home/<username>/.config/chromium/Default/Local Storage/leveldb')
local_storage_value = db.get(b'_https://app.slack.com\x00\x01localConfig_v2')
slack_config = json.loads(local_storage_value[1:])
```

This requires the [Plyvel](https://plyvel.readthedocs.io/en/latest/) library which we can conditionally import when running for Chrome like Pycryptodome. Note that Chrome has to be closed it to be able to read the library. We should gracefully handle the error when Chrome is open.